# Classification Template
In this code I will analyse which is the best classification algorithm to predict a benign/malign cancer tumor. 
You can find the database using the sklearn databases or accesing through the UCI ML Repository. 
Breast Cancer Wisconsin (Diagnostic) datasets.
https://goo.gl/U2Uwz2
The dataset contains 569 observations of patients with 30 numeric attributes and kind of tumor.
Here, I will take only 75% of the sample to train the algorithms and then compare among them which one makes the best predictions.
The algorithms used are:


*   Random Forest Classification
*   Decision Tree Classification 
* Linear SVM
* Kernel SVM
* Naive Bayes
* Logistic Regression

After getting the confusion matrix I will give my conclusions on why I consider one might be performing better than the other.





## Importing the libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [ ]:
from sklearn.datasets import load_breast_cancer
cancer = load_breast_cancer()
#cancer.data

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [ ]:
print(cancer.keys())
print(cancer.target_names)
print(cancer.target)

print(cancer.feature_names)
# INFO ABOUT THE DATASET 
#print(cancer.DESCR)


dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])
['malignant' 'benign']
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1

In [ ]:
#creating a data frame
df = pd.DataFrame(np.c_[cancer['data'], cancer['target']],
                  columns = np.append(cancer['feature_names'], ['typeoftumor']))
df.tail()
X = pd.DataFrame(df.iloc[:,:-1].values)
y = pd.DataFrame(df.iloc[:,-1].values)

X.head()

0      1       2       3   ...      26      27      28       29
0  17.99  10.38  122.80  1001.0  ...  0.7119  0.2654  0.4601  0.11890
1  20.57  17.77  132.90  1326.0  ...  0.2416  0.1860  0.2750  0.08902
2  19.69  21.25  130.00  1203.0  ...  0.4504  0.2430  0.3613  0.08758
3  11.42  20.38   77.58   386.1  ...  0.6869  0.2575  0.6638  0.17300
4  20.29  14.34  135.10  1297.0  ...  0.4000  0.1625  0.2364  0.07678

[5 rows x 30 columns]

## Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

#indicates how many items are in the array
print(X_train.shape, y_train.shape)
#the train set contain 426
print(X_test.shape, y_test.shape)
#the test 143

(426, 30) (426, 1)
(143, 30) (143, 1)


## Feature Scaling
Since many of the features are around 0 and 2 and some other greater than 100 and others greater thatn 1000, I decided to run feature scaling to avoid biases due to these differences in scale

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

#KNearest Neighbors

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
knn.fit(X_train, y_train)

#predict
y_pred=knn.predict(X_test)
#print(np.concatenate((y_test.reshape(len(y_test),30),y_pred.reshape(len(y_pred),30)),30))

#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))



[[47  6]
 [ 1 89]]
0.951048951048951
              precision    recall  f1-score   support

         0.0       0.98      0.89      0.93        53
         1.0       0.94      0.99      0.96        90

    accuracy                           0.95       143
   macro avg       0.96      0.94      0.95       143
weighted avg       0.95      0.95      0.95       143



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


## Random Forest Classification model

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rdm = RandomForestClassifier(n_estimators=100, criterion='entropy', random_state=0)
#n_estimators= number of trees, maybe with more trees it would overfit
rdm.fit(X_train,y_train)

#predict
y_pred=rdm.predict(X_test)


#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  after removing the cwd from sys.path.


[[52  1]
 [ 3 87]]
0.972027972027972
              precision    recall  f1-score   support

         0.0       0.95      0.98      0.96        53
         1.0       0.99      0.97      0.98        90

    accuracy                           0.97       143
   macro avg       0.97      0.97      0.97       143
weighted avg       0.97      0.97      0.97       143



## Support Vector Machine SVM (linear)

In [ ]:
from sklearn.svm import SVC
lsvm = SVC( kernel='linear', random_state=0)
lsvm.fit(X_train,y_train)

#predict
y_pred=lsvm.predict(X_test)


#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))




[[51  2]
 [ 2 88]]
0.972027972027972
              precision    recall  f1-score   support

         0.0       0.96      0.96      0.96        53
         1.0       0.98      0.98      0.98        90

    accuracy                           0.97       143
   macro avg       0.97      0.97      0.97       143
weighted avg       0.97      0.97      0.97       143



/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


##Kernel SVM

In [ ]:
from sklearn.svm import SVC
#use and rbf kernel instead a linear . this is what 
ksvm = SVC(kernel = 'rbf', random_state = 0)
ksvm.fit(X_train, y_train)

#predict
y_pred=ksvm.predict(X_test)


#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))



[[50  3]
 [ 2 88]]
0.965034965034965
              precision    recall  f1-score   support

         0.0       0.96      0.94      0.95        53
         1.0       0.97      0.98      0.97        90

    accuracy                           0.97       143
   macro avg       0.96      0.96      0.96       143
weighted avg       0.96      0.97      0.96       143



/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


## Naive Bayes

In [ ]:
from sklearn.naive_bayes import GaussianNB
gauss = GaussianNB()
gauss.fit(X_train,y_train)

#predict
y_pred=gauss.predict(X_test)


#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[47  6]
 [ 6 84]]
0.916083916083916
              precision    recall  f1-score   support

         0.0       0.89      0.89      0.89        53
         1.0       0.93      0.93      0.93        90

    accuracy                           0.92       143
   macro avg       0.91      0.91      0.91       143
weighted avg       0.92      0.92      0.92       143



/usr/local/lib/python3.6/dist-packages/sklearn/naive_bayes.py:206: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


##Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
logistic = LogisticRegression(random_state=0)
logistic.fit(X_train,y_train)

#predict
y_pred=logistic.predict(X_test)


#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[50  3]
 [ 3 87]]
0.958041958041958
              precision    recall  f1-score   support

         0.0       0.94      0.94      0.94        53
         1.0       0.97      0.97      0.97        90

    accuracy                           0.96       143
   macro avg       0.96      0.96      0.96       143
weighted avg       0.96      0.96      0.96       143



/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


##Conclusion
As you could see, according to the accuracy tests, I will list from the best to the least performer.

1.   Random Forest
2.   SVM Linear
3. Kernel SVM
4. Logistic 
5. KNN
6. Naive Bayes

Here it is easy to see that the worst performer is Naive Bayes with 91%, I attribute this to the fact that in this case most likely many of the variables are correlated. Thus, since the main assumption of Bayes is that the variables are independent in this case it might no be a plausible assumption, therefore its bad performance.
It was even beaten by a linear model like the logistic one.

As the best ones are the Random Forest and the linear Support Vector Machine with a 97% of accuracy, nevertheless it is iportant to see which were the best predicted. In both cases, they predict better the 1's (Benign tumors). In comparison, the KNN with 95% of accuracy, predicts with higher accuracy the 0's, which stand for Malign tumors, therefore, I can only conclude that depending on the goal of the client, in this case, maybe a doctor, a model with less accuracy does not mean it is the worst. 

At the end, it depends what is more important to predict, whether the model can indicate which ones are Benign or Malign.






